### ISSUE TITLE
trtri undefined references in cuda/11.2.2 build with no eti

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos-kernels#2231]
Builds (at least with cuda/11.2.2) with no eti fail to link trtri:

```
17:34:43 [ 63%] Linking CXX executable KokkosKernels_lapack_serial
17:34:45 CMakeFiles/KokkosKernels_lapack_serial.dir/backends/Test_Serial_Lapack.cpp.o: In function `int KokkosLapack::trtri<Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Serial> >(char const*, char const*, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Serial> const&)':
17:34:45 tmpxft_00007941_00000000-6_Test_Serial_Lapack.cudafe1.cpp:(.text._ZN12KokkosLapack5trtriIN6Kokkos4ViewIPPdJNS1_10LayoutLeftENS1_6SerialEEEEEEiPKcS9_RKT_[_ZN12KokkosLapack5trtriIN6Kokkos4ViewIPPdJNS1_10LayoutLeftENS1_6SerialEEEEEEiPKcS9_RKT_]+0x460): undefined reference to `KokkosLapack::Impl::TRTRI<Kokkos::View<int, Kokkos::LayoutRight, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> >, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, false, false>::trtri(Kokkos::View<int, Kokkos::LayoutRight, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > const&, char const*, char const*, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&)'
17:34:45 CMakeFiles/KokkosKernels_lapack_serial.dir/backends/Test_Serial_Lapack.cpp.o: In function `int KokkosLapack::trtri<Kokkos::View<Kokkos::complex<double>**, Kokkos::LayoutLeft, Kokkos::Serial> >(char const*, char const*, Kokkos::View<Kokkos::complex<double>**, Kokkos::LayoutLeft, Kokkos::Serial> const&)':
17:34:45 tmpxft_00007941_00000000-6_Test_Serial_Lapack.cudafe1.cpp:(.text._ZN12KokkosLapack5trtriIN6Kokkos4ViewIPPNS1_7complexIdEEJNS1_10LayoutLeftENS1_6SerialEEEEEEiPKcSB_RKT_[_ZN12KokkosLapack5trtriIN6Kokkos4ViewIPPNS1_7complexIdEEJNS1_10LayoutLeftENS1_6SerialEEEEEEiPKcSB_RKT_]+0x460): undefined reference to `KokkosLapack::Impl::TRTRI<Kokkos::View<int, Kokkos::LayoutRight, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> >, Kokkos::View<Kokkos::complex<double>**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, false, false>::trtri(Kokkos::View<int, Kokkos::LayoutRight, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > const&, char const*, char const*, Kokkos::View<Kokkos::complex<double>**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&)'
17:34:45 collect2: error: ld returned 1 exit status
```

**Reproducer**: (weaver rhel8)
```bash
bsub -Is -n 1 -q rhel8 -gpu "num=1" bash

source /etc/profile.d/modules.sh
source /projects/ppc64le-pwr9-rhel8/legacy-env.sh

module load cuda/11.2.2/gcc/8.3.1 cmake/3.23.1

${KOKKOSKERNELS_PATH}/cm_generate_makefile.bash --with-cuda --with-serial --compiler=${KOKKOS_PATH}/bin/nvcc_wrapper --arch=Volta70,Power9 --with-cuda-options=enable_lambda --kokkos-path=${KOKKOS_PATH} --kokkoskernels-path=${KOKKOSKERNELS_PATH} --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --cxxstandard=17 --no-default-eti

make -j16
```

**Comment by cwpearson** [comment]
I think the problem is a combination of three things

here we try to build the `KokkosKernels_lapack_serial` test whenever the `Serial` backend is enabled

https://github.com/kokkos/kokkos-kernels/blob/b2210058826672c8de838541a36f7b946ecbb79a/lapack/unit_test/CMakeLists.txt#L65-L73

Here we declare `template <..., false> struct TRTRI` but only define it when either
* `!defined(KOKKOSKERNELS_ETI_ONLY)`
* `KOKKOSKERNELS_IMPL_COMPILE_LIBRARY`

https://github.com/kokkos/kokkos-kernels/blob/b2210058826672c8de838541a36f7b946ecbb79a/lapack/impl/KokkosLapack_trtri_spec.hpp#L61-L71

As expected, in every generated ETI file, `KOKKOSKERNELS_IMPL_COMPILE_LIBRARY` is `true`, so that defines `template <..., true> struct TRTRI`, but not the missing `template <..., false> struct TRTRI`

However, when we compile the serial test file KOKKOSKERNELS_ETI_ONLY is defined and `KOKKOSKERNELS_IMPL_COMPILE_LIBRARY` is not, which means we're trying to use `template <..., false> struct TRTRI` which is never declared

Test_Serial.hpp has

```c++
#if defined(KOKKOSKERNELS_TEST_ETI_ONLY) && !defined(KOKKOSKERNELS_ETI_ONLY)
#define KOKKOSKERNELS_ETI_ONLY
#endif
```

cm_test_all_sandia sets `ENABLE_TEST_ETI_ONLY=True`, which means `-DKokkosKernels_TEST_ETI_ONLY=OFF` is never set

So I think the problem is:
1. This configuration does not ETI serial
2. We build the serial test
3. This configuration does not set KokkosKernels_TEST_ETI_ONLY=OFF
4. KokkosKernels_TEST_ETI_ONLY defaults to ON
5. When TEST_ETI_ONLY is on, of course the serial test then cannot find the non-ETI'ed Serial functions.

**Comment by ndellingwood** [comment]
@lucbv I retested this with d332d919644c07320510389ed503072e1b13fd99 and still seeing linker errors:
```
CMakeFiles/KokkosKernels_lapack_serial.dir/backends/Test_Serial_Lapack.cpp.o: In function `int KokkosLapack::trtri<Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Serial> >(char const*, char const*, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Serial> const&)':
tmpxft_003edc87_00000000-6_Test_Serial_Lapack.cudafe1.cpp:(.text._ZN12KokkosLapack5trtriIN6Kokkos4ViewIPPdJNS1_10LayoutLeftENS1_6SerialEEEEEEiPKcS9_RKT_[_ZN12KokkosLapack5trtriIN6Kokkos4ViewIPPdJNS1_10LayoutLeftENS1_6SerialEEEEEEiPKcS9_RKT_]+0x36c): undefined reference to `KokkosLapack::Impl::TRTRI<Kokkos::View<int, Kokkos::LayoutRight, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> >, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, false, false>::trtri(Kokkos::View<int, Kokkos::LayoutRight, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > const&, char const*, char const*, Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&)'
CMakeFiles/KokkosKernels_lapack_serial.dir/backends/Test_Serial_Lapack.cpp.o: In function `int KokkosLapack::trtri<Kokkos::View<Kokkos::complex<double>**, Kokkos::LayoutLeft, Kokkos::Serial> >(char const*, char const*, Kokkos::View<Kokkos::complex<double>**, Kokkos::LayoutLeft, Kokkos::Serial> const&)':
tmpxft_003edc87_00000000-6_Test_Serial_Lapack.cudafe1.cpp:(.text._ZN12KokkosLapack5trtriIN6Kokkos4ViewIPPNS1_7complexIdEEJNS1_10LayoutLeftENS1_6SerialEEEEEEiPKcSB_RKT_[_ZN12KokkosLapack5trtriIN6Kokkos4ViewIPPNS1_7complexIdEEJNS1_10LayoutLeftENS1_6SerialEEEEEEiPKcSB_RKT_]+0x36c): undefined reference to `KokkosLapack::Impl::TRTRI<Kokkos::View<int, Kokkos::LayoutRight, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> >, Kokkos::View<Kokkos::complex<double>**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, false, false>::trtri(Kokkos::View<int, Kokkos::LayoutRight, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > const&, char const*, char const*, Kokkos::View<Kokkos::complex<double>**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [lapack/unit_test/CMakeFiles/KokkosKernels_lapack_serial.dir/build.make:125: lapack/unit_test/KokkosKernels_lapack_serial] Error 1
make[1]: *** [CMakeFiles/Makefile2:2465: lapack/unit_test/CMakeFiles/KokkosKernels_lapack_serial.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs...
```

Updated reproducer (blake H100 queue):
```
module load cmake cuda/12.8.0 gcc/12.2.0

$KOKKOSKERNELS_PATH/cm_generate_makefile.bash --with-cuda --with-serial --compiler=${KOKKOS_PATH}/bin/nvcc_wrapper --arch=HOPPER90 --kokkos-path=${KOKKOS_PATH} --kokkoskernels-path=${KOKKOSKERNELS_PATH} --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --cxxstandard=20 --no-default-eti

make -j20
```

**Comment by lucbv** [comment]
After some reading of our ETI system I think I have a decent picture of what is going here. Of course @cwpearson is right that the link error is due to a lack of ETI when our unit-tests are assuming it. The root cause is the non-uniformity of the logic around the instantiation of execution spaces, memory spaces and `HostSpace` which does not follow either of the former.
Let's review the instantiation logic for these three "types"

## TLDR

`HostSpace` default instantiation follows the value of ${KOKKOSKERNELS_ADD_DEFAULT_ETI}, instead of checking if one of `Kokkos::Serial`, `Kokkos::OpenMP` or `Kokkos::Threads` is instantiated, which is the logic for setting the default instantiation of device memory spaces.

## Execution Spaces

That one starts with [kokkos_backends.cmake](https://github.com/kokkos/kokkos-kernels/blob/develop/cmake/kokkos_backends.cmake), there we define the default value for execution spaces instantiation to be the same as `Kokkos_ENABLE_{BackEnd}` so we inherit the default value from Kokkos Core's configuration. That default value is then used in [kokkoskernels_eti_devices.cmake](https://github.com/kokkos/kokkos-kernels/blob/develop/cmake/kokkoskernels_eti_devices.cmake) to define the actual instantiation of the execution space as follows:
```
kokkoskernels_add_option("INST_EXECSPACE_{BackEnd}" ${KOKKOSKERNELS_INST_EXECSPACE_{BackEnd}_DEFAULT} BOOL "...")
```
While the `CUDA` execution space is defined slightly differently, it eventually follows the same logic.

## Memory Spaces

If entering the `KOKKOS_ENABLE_CUDA` block, `CUDASPACE` instantiation is `ON` by default and `CUDAUVMSPACE` instantiation is `OFF` by default.
If entering the `KOKKOS_ENABLE_HIP` block, `HIPSPACE` instantiation is `${KOKKOSKERNELS_INST_EXECSPACE_HIP_DEFAULT}` by default and `HIPMANAGEDSPACE` instantiation is `OFF` by default.
SYCL follows HIP's logic.

`INST_MEMSPACE_HOSTSPACE` is the outlier being defined as `${KOKKOSKERNELS_ADD_DEFAULT_ETI}` by default.

## Devices 

Finally, two lists are constructed, `DEVICE_LIST` and `DEVICES`, both follow the same pattern of checking that both the execution space and memory space that form the device are to be instantiated.
